File: 103-binary-tree-zigzag-level-order-traversal/103-binary-tree-zigzag-level-order-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> zigZag;
        if(!root) return zigZag;
        
        bool leftToRight = true;
        queue<TreeNode*> q;
        q.push(root);
        
        while(!q.empty()) {
            int n = q.size();
            vector<int> level(n);
            
            for(int i = 0; i < n; ++i) {
                TreeNode* current = q.front();
                q.pop();
                
                // find the position to insert the node
                int idx = leftToRight ? i : n - i - 1;
                level[idx] = current->val;
                
                if(current->left) q.push(current->left);
                if(current->right) q.push(current->right);
            }
            
            leftToRight = !leftToRight;
            zigZag.emplace_back(level);
        }
        
        return zigZag;
    }
};

File: 103-binary-tree-zigzag-level-order-traversal/103-binary-tree-zigzag-level-order-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from queue import Queue
class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        zigZag = []
        if not root: return zigZag
        q = Queue()
        q.put(root)
        leftToRight = True
        while q.qsize():
            n = q.qsize()
            level = [0] * (n)
            for i in range (n):
                current = q.get()
                idx = i if leftToRight else n - i - 1
                level[idx] = current.val
                if current.left: q.put(current.left)
                if current.right: q.put(current.right)
            leftToRight = not leftToRight
            zigZag.append(level)
        return zigZag